### PR TITLE
De-duplicate link fetching to one request at once

### LINKED ISSUE #1614: Await existing link expander promises
Link previewer calls `fetch` for every url in every message, it may happen that one url is posted multiple times (in different channels), and before the response arrives for the first request, there is no real need to fire another request, something like
```
if (requests[link]) {
    requests[link].then()
}
```

https://github.com/thelounge/lounge/blob/f18d0193837cac829c26cfc90361752de91f9164/src/plugins/irc-events/link.js#L38-L46

### PR DESCRIPTION
**Pull request by xPaw**
Converts `fetch` into a Promise, and returns existing Promise if there's one already.

Hashed by `uri` and `headers` passed in so that language is considered.

Supercedes #1952.
Fixes #1614.

### CODE REVIEW

File: src/plugins/irc-events/link.js [hunk 1]
```diff
@@ -313,71 +310,89 @@ function getRequestHeaders(headers) {
 	return formattedHeaders;
 }
 
-function fetch(uri, headers, cb) {
-	let req;
+function fetch(uri, headers) {
+	// Stringify the object otherwise the objects won't compute to the same value
+	const cacheKey = JSON.stringify({uri, headers});
```
Comment (xPaw) [inline src/plugins/irc-events/link.js]: Probably should be an array instead of an object to reduce length of the string.

Comment (astorije) [reply src/plugins/irc-events/link.js]: That would go from:
`'{"uri":"https://example.com","headers":{"foo":"bar"}}'`
to:
`'["https://example.com",{"foo":"bar"}]'`
Correct?

If so, that saves us 16 characters per string, I wouldn't worry too much.
I'm obviously not opposed if you prefer to do that though.

File: src/plugins/irc-events/link.js [hunk 2]
```diff
@@ -313,71 +310,89 @@ function getRequestHeaders(headers) {
...
-	} catch (e) {
-		return cb(null, e);
+	if (promise) {
+		return promise;
 	}
 
-	const buffers = [];
-	let length = 0;
-	let limit = Helper.config.prefetchMaxImageSize * 1024;
+	const removeCache = () => currentFetchPromises.delete(cacheKey);
```
Comment (astorije) [inline src/plugins/irc-events/link.js]: I'm probably missing something, but any reason why you're creating this here and not just above where it's used, L391?

File: test/plugins/link.js [hunk 3]
```diff
@@ -422,4 +422,40 @@ describe("Link plugin", function() {
 
 		expect(message.previews).to.be.empty;
 	});
+
+	it("should fetch same link only once at the same time", function(done) {
```
Comment (astorije) [inline test/plugins/link.js]: 👏

File: src/plugins/irc-events/link.js (hunk 2 above)
Comment (xPaw) [reply src/plugins/irc-events/link.js]: I tried chained promises, forgot to move it back down.

File: src/plugins/irc-events/link.js (hunk 1 above)
Comment (xPaw) [reply src/plugins/irc-events/link.js]: Worth it!

**Pull request merged**